core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -251,13 +253,25 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
               type.getFieldNames()),
           type.isNullable());
     } else if (type instanceof JavaType) {
-      return typeFactory.createTypeWithNullability(
-          typeFactory.createSqlType(type.getSqlTypeName()),
-          type.isNullable());
+      Class javaClass = ((JavaType) type).getJavaClass();
+      final RelDataType nonNullType;
+      if (javaClass.isArray()) {
```

Comment:
what about map type?

Reply:
Issue described in Jira ticket relates to ```SqlTypeFactoryImpl```'s assertBasic method: 
```java
  private void assertBasic(SqlTypeName typeName) {
    assert typeName != null;
    assert typeName != SqlTypeName.MULTISET
        : "use createMultisetType() instead";
    assert typeName != SqlTypeName.ARRAY
        : "use createArrayType() instead";
    assert typeName != SqlTypeName.ROW
        : "use createStructType() instead";
    assert !SqlTypeName.INTERVAL_TYPES.contains(typeName)
        : "use createSqlIntervalType() instead";
  }
```
For map type assertion isn't fired here and I think it's rare case that someone will accept map arguments (or others from the method body) into custom table function.

Reply:
I'm not sure that's a bug or is just designed.
I create a PR(https://github.com/apache/calcite/pull/1521) trying to fix it